basic grid autogenerate logic

jbusecke opened the pull request:
Adds basic functionality to create 'c-grid style' coordinates to xarray datasets (e.g. observational datasets)

rabernat commented:
Wow this is great! I will have a close look tomorrow.

jbusecke commented:
OK @rabernat, this should look a bit better than the first commit! Excited to hear what you think.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,254 @@
+                    elif to is 'right':
+                        fill_value = bot
+                else:
+                    fill_value = z_pad
```
I'm not sure if the design of this function is optimal. Could we not just have a generic `generate_axis` function that has sufficient flexibility via kwargs to apply to any scenario? Then we would loop over axes in `autogenerate_ds` and call this `generate_axis` function on each.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,254 @@
+        The position of the datapoints relative to the coordinates given
+        This will usually be the cell center, but can sometimes be the
+        lower left (option:'left') or perhaps upper right (option:'right')
+        corner of the gridcell.
```
Shouldn't this be axis dependent? Could one specific a dict for position like `{'X': 'center', 'Y': 'right', 'Z' left'}`?

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,254 @@
+        data
+    x_coord_wrap : float - Like 'x_wrap' but for coordinate variable
+    y_coord_wrap : float - Like 'y_wrap' but for coordinate variable
+    z_coord_pad : float -  Like 'z_pad' but for coordinate variable
```
Again, I am opposed to hard-coding the possible axes. The design should be more general and less redundant.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,254 @@
+
+    # Loop over all specified axes and build new coordinates
+    ds = generate(ds, axes, axes, to, x_wrap, y_wrap,
+                  z_pad, to_shift, from_shift, dim_switch=True)
```
I don't see a loop. Why do you say "loop over all specified axes"? Maybe that loop should be in this function, not `generate`...

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,254 @@
+    # 'coord_axes'
+    if coord_axes:
+        ds = generate(ds, coord_axes, axes, to, x_coord_wrap, y_coord_wrap,
+                      z_coord_pad, to_shift, from_shift, dim_switch=False)
```
It seems weird to be calling `generate` again. Can you explain what is going on here?

rabernat commented on xgcm/grid.py:
```diff
@@ -555,6 +569,15 @@ def cumsum(self, da, axis, **kwargs):
+    t_order = list(range(len(da.data.shape)))
+    t_order[axis] = 0
+    t_order[0] = axis
+    data = np.transpose(da.data.copy(), axes=t_order)
```
Does calling transpose on dask data trigger computation? If so, this is forbidden.

rabernat replied:
Can you explain what this function does?

jbusecke replied:
This function was my attempt to enable the 'wrapping' of values around a discontinuity. The basic idea is to add a 'wrap value' to the first or last (depending on the shift_to direction) slice from an arbitrary axis (xarray dimension). Is there a way to selectively replace values in a dask array without triggering computation?

Alternatively we might have to compute and then return to a dask array?

rabernat commented on xgcm/test/test_autogenerate.py:
```diff
@@ -0,0 +1,152 @@
+
+    xr.testing.assert_identical(center, ds_out_center)
+    xr.testing.assert_identical(left, ds_out_left)
+    xr.testing.assert_identical(right, ds_out_right)
```
I really like your test!

It is, however, a pretty high level test. It might be valuable to break down your new module into smaller units and write some unit tests.

rabernat commented on xgcm/test/test_grid.py:
```diff
@@ -432,3 +432,16 @@ def test_grid_ops(all_datasets):
+                    [1., 4., 1.]],
+                    [[1., 4., 1.],
+                    [1., 4., 1.]]])
+    np.testing.assert_allclose(new, ref)
```
Can you test this function on dask data and ensure that it does not trigger computation?

rabernat commented:
We probably could have discussed the API and design a bit more before diving in, but I'm really glad you got this started.

Here's how I would have suggested approaching it.

- Put all the heavy lifting into a function called something like `autogenerate_axis_coords` which operates on a single axis at a time. It would have lots of options (boundary conditions, etc.). It would return a new dataset with extra coords in it.
- Have a simple function called `autogenerate_grid_coords` which just loops over the axes and calls `autogenerate_axis_coords` for each one.
- Have convenience methods called `autogenerate_grid` and `autogenerate_coords` which call the above methods and then make `Axis` and `Grid` objects out of them.

jbusecke commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,254 @@
+                The differenced data
+            """
+            #TODO I think we can modify xgcm.Axis so that this can all be
+            # absorbed in there.
```
I think we can achieve this functionality (which is mostly copied and pasted) by modifying the xgcm.Axes object, then we can completely remove this from here.

rabernat replied:
I would avoid modifying the Axis class. This can all be accomplished with a simple function. No classes needed. (See my comment below).

jbusecke commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,254 @@
+                                              boundary=boundary,
+                                              fill_value=fill_value)
+
+            # Is there a way to use the existing grid.Axis.interp logic?
```
Even more reason to absorb into xgcm.Axis

jbusecke replied:
Oh sorry I guess I only read the last part of the comment. Ill see what I can do!

jbusecke commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,254 @@
+        360 for global longitude)
+
+        y_wrap - float: as x_wrap but for y direction (defaults to
+        180 for global longitude)
```
My intention was to make this work relatively smoothly "out of the box with common datasets (like satellite products). The logic should work with any succession of values that need to be wrapped around a discontinuity. This would have to be explained better in the docstring, but if you have an arbitrary dataset with x coordinates 1,2,...10 you would have to specify 'x_wrap=10' and it should do the right thing. Should have a test too.

jbusecke commented:
Thanks for the detailed input, Ryan. I agree that this can all be made much more general, but I just wanted to start somewhere and get something that achieves the basics. I think a lot of the hard coding can be removed with proper 'bookeeping' of the kwargs.

Ill get working on that on the train. Yay, commuting!

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -1,254 +1,240 @@
+        Dataset with gridinformation used to construct c-grid
+    axis_dict : dict
+        Dict with information on the dimension in ds corrsponding to the xgcm
+        axis. E.g. {'X':'lon'}
```
I don't see `axis_dict` in the function arguments

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -1,254 +1,240 @@
+def generate_axis(ds,
+                  axis,
+                  name,
+                  axis_dim,
```
The docstring is not consistent with the function arguments

jbusecke replied:
Fixed it.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -1,254 +1,240 @@
+    axis_dict : dict
+        Dict with information on the dimension in ds corrsponding to the xgcm
+        axis. E.g. {'X':'lon'}
+    pos_from : str
```
should be `pos_from : {'center','left','right'}`

jbusecke commented:
Ok I rewrote the module substantially. I think I at least started addressing most of your comments. I have still a bunch of cleaning up to do, but I wanted to get some feedback first.
A typical usecase would be this:
```python
from xgcm import Axis, Grid
import xarray as xr
import numpy as np
from dask.array import from_array
from xgcm.autogenerate import autogenerate_grid_ds

# create test datasets
dx = 0.5
dy = 1.0
dz = 0.5
a = np.random.rand(180, int(360/dx), int(10/dz))

x = np.arange(-180, 180, dx)
y = np.arange(-90, 90, dy)
z = np.arange(0, 10, dz)

xx, yy = np.meshgrid(x, y)
_, _, zz = np.meshgrid(x, y, z)

ds_original = xr.Dataset(
                         {'somedata': (['lat', 'lon', 'z'], a)},
                         coords={'lon': (['lon', ], x+(dx/2.0)),
                                 'lat': (['lat', ], y+(dy/2.0)),
                                 'z': (['z', ], z+(dx/2.0)),
                                 'llon': (['lat', 'lon'], xx+(dx/2.0)),
                                 'llat': (['lat', 'lon'], yy+(dy/2.0)),
                                 'zz': (['lat', 'lon', 'z'], zz+(dx/2.0))}
                        )

axis_dims =  {'X':'lon','Y':'lat','Z':'z'}
axis_coords =  {'X':'llon','Y':'llat','Z':'zz'}
ds_new = autogenerate_grid_ds(ds_original,
                              axis_dims,
                              axis_coords,
                              wrap={'lon':360, 'lat':180,
                                    'llon':360, 'llat':180},
                              pad={'z':'auto','zz':'auto'})
ds_new
```

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -1,254 +1,240 @@
+    if raw_switch:
+        # Input coordinate has to be declared as center,
+        # or xgcm.Axis throws error. Will be rewrapped below.
+        ds[name] = fill_attrs(ds[name], 'center', axis)
```
I would probably copy `ds` before adding new variables to it

jbusecke replied:
Good call, that caused some issues in the tests earlier. Added `ds=ds.copy` before

rabernat commented on xgcm/grid.py:
```diff
@@ -227,13 +227,17 @@ def _neighbor_binary_func(self, da, f, to, boundary=None, fill_value=0.0):
         # get the two neighboring sets of raw data
-        data_left, data_right = self._get_neighbor_data_pairs(da, to,
-                                      boundary=boundary, fill_value=fill_value)
+        data_left, data_right = self._get_neighbor_data_pairs(da, to,\
```
please take this `\` out

jbusecke replied:
Done.

rabernat commented on xgcm/grid.py:
```diff
@@ -74,6 +73,7 @@ def __init__(self, ds, axis_name, periodic=True, default_shifts={},
         self._name = axis_name
         self._periodic = periodic
         self._wrap = wrap
+        self._raw_out = raw_out
```
What is this attribute?!?

jbusecke replied:
This is basically a switch to omit the call to `_wrap_and_replace_coords` at the end. This way I can use all the logic from `xgcm.Axis` to do the interpolation (as mentioned above). I think this is much better then keeping a branched out interp function in the autogenerate module.

rabernat commented on xgcm/grid.py:
```diff
@@ -569,14 +573,21 @@ def cumsum(self, da, axis, **kwargs):
+    # beginning or end before or after will be empty)
+    before = da[{dim: slice(0, sl)}]
+    middle = da[{dim: sl}]
+    after = da[{dim: slice(sl+1, None)}]
```
perfect...this should be dask friendly

jbusecke replied:
is there a preferred way to test if a computation was triggered?

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,240 @@
+                  pad=None,
+                  raw_switch=True):
+    """
+    Parameters
```
We need a description of what the function does.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,240 @@
+        enabled for dimensions and deactivated for multidimensional
+        coordinates. These can only be calculated after the dims are created.
+    """
+    new_name = name+'_inferred'
```
this suffix should probably be a keyword argument

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,240 @@
+    else:
+        ax = Axis(ds, axis, periodic=periodic, wrap=wrap, raw_out=False)
+        ds.coords[new_name] = ax.interp(ds[name], pos_to, boundary=boundary,
+                                        fill_value=fill_value)
```
Can you help me understand what is happening here? What is `raw_switch`? What do these two cases represent?

jbusecke replied:
Ok so for a set of coordinates there are two fundamental cases. The coordinates are a) one dimensional (dimensions) or 2) multidimensional. These are separated by the keyword  raw_switch.

These two cases are treated differently because for each dataset we need to recreate all a) cases before we can proceed to 2), hence this is really the 'raw' data processing step. If we have working one dimensional coordinates (e.g. after we looped over the `axes_dims_dict`, we can use the regular xgcm.Axis to interpolate multidimensional coordinates. 
I designed it this way so that any changes to the Axis.interp method can directly propagate to this module.

Does that make any sense?

rabernat replied:
Now I understand and it makes sense.

That would have been a good comment to include in the code.

jbusecke replied:
It is included now

jbusecke commented:
Just pushed a commit with fixes requested above.

jbusecke commented:
@rabernat, could you take a look at why the python 2 test is failing? Is there an easy way to fix that? Can I set the broadcasting rule like in python 3?

jbusecke commented:
I will work on getting the test coverage up.

jbusecke commented:
@rabernat, should I clean and finish this part as is right now or also try to integrate some more logic, to calculate distances, areas, and volumes? This would be complimentary to development in #75. I think we should probably keep that seperate and finish this now. Just want to hear what you think.

rabernat commented:
An alternative instead of the `raw_switch` approach.

Rewrite `Grid._neighbor_binary_ufunc` as
```
    # returns numpy or dask array
    data_new = self._binary_ufunc_raw(...)
    return self._wrap_and_replace_coords(da, data_new, to)
```

And then in your code, you call `_binary_ufunc_raw`. This function would need to be defined in `Grid`.

rabernat commented on xgcm/grid.py:
```diff
@@ -284,12 +293,22 @@ def _get_neighbor_data_pairs(self, da, position_to, boundary=None,
+                left = da.roll(**{dim: 1})
+                left = add_to_slice(left, dim, 0, -self._wrap).data
+            else:
+                left = da.roll(**{dim: 1}).data
```
can we put this wrap logic in standalone function?

jbusecke replied:
I think yes, but I would probably have to repeat all the positional logic. I did sinplify the wrap logic, soo maybe that is acceptable?

rabernat commented on xgcm/grid.py:
```diff
@@ -284,12 +293,22 @@ def _get_neighbor_data_pairs(self, da, position_to, boundary=None,
+                right = da.roll(**{dim: -1})
+                right = add_to_slice(right, dim, -1, self._wrap).data
+            else:
+                right = da.roll(**{dim: -1}).data
```
Ditto

jbusecke commented:
I am not able to “access” the basic logic in ‘interp_function’ within Axis.interp. Can You advise me how to pass the inner function as an input?

jbusecke commented:
Ok I added tests for everything and removed the 'raw' switch.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -224,6 +234,13 @@ def position_to_relative(pos_from, pos_to):
+    # TODO: This is not great, but I am not sure how to pass the nested
+    # function in xgcm.grid.Axis.interp to Axis._neighbor_binary_func_raw
+    # instead.
+    return 0.5*(data_left + data_right)
```
You are correct that you can't extract that function because it is defined within a another. But there is a simple solution: define it at the top level of the grid module namespace and then use the same function in your routine and `interp`.

What I mean is to just define
```python
def interp_raw(data_left, data_right):
    # linear, centered interpolation
    # TODO: generalize to higher order interpolation
    return 0.5*(data_left + data_right)

def diff_raw ...
```
in grid.py, change the `diff` and `interp` methods to call these functions, and then import the same `_raw` functions in your module.

Does this make sense to you?

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,274 @@
+    raw_switch : bool
+        Determines if the attributes are created from scratch. Should be
+        enabled for dimensions and deactivated for multidimensional
+        coordinates. These can only be calculated after the dims are created.
```
I would still prefer a different name for `raw_switch`. Why not `create_attributes_from_scratch`? Something more verbose that corresponds with what the option actually does.

rabernat commented:
Still need:

- [ ] documentation (like a new page called `doc/grid_generation.rst` and / or an example notebook modeled on `/doc/example_mitgcm.ipynb`)
- [ ] fix the python 2.7 test failure

jbusecke commented:
Regarding the python 2.7 test failure: It seems like some numpy functions behave differently (e.g. the assert_isclose() function). Is there a simple way to import the python 3 behaviour as for the print statement?

rabernat commented:
> Regarding the python 2.7 test failure: It seems like some numpy functions behave differently (e.g. the assert_isclose() function). Is there a simple way to import the python 3 behaviour as for the print statement?

You're gonna have to sort this out on your own.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -272,3 +285,28 @@ def fill_attrs(da, pos, axis):
+    dx = dlon * np.cos(np.deg2rad(lat.data)) * dll_factor
+    dy = dlat * dll_factor
+
+    return dx, dy
```
This function looks useful, but why is it in this PR? It seems like a new, separate feature.

jbusecke commented:
Seems like the Travis build is passing now. While writing the docs I decided to implement the distance calculation at this point already. This requires a few more functions and mostly tests. Should be ready soon.

rabernat commented:
> While writing the docs I decided to implement the distance calculation at this point already. 
Not in this PR. We need to discuss the design and API for that stuff separately.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,312 @@
+     Name prefix for distance coordinates. Defaults to 'd'+name.
+    ll_dist : Bool
+     Activates the conversion of lon/lat coordinates to distances in meters.
+     Requires coordinates to be specified on 'X' and 'Y'.
```
It doesn't look like this option is ever utilized.

I'm confused about what it would mean. How can you covert lon/lat coordinates to distance without specifying a projection?

rabernat commented:
Just to be clear...I am very very happy with this PR :clap: :clap: :clap:!

I just don't want to clutter it up with new features just as we are getting ready to merge.

jbusecke commented:
Ahh shoot. This was a big git fail...I am moving these new features into a new branch where they should have gone in the first place. Need to finish the docs still.

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,272 @@
+    attrs_from_scratch : bool
+        Determines if the attributes are created from scratch. Should be
+        enabled for dimensions and deactivated for multidimensional
+        coordinates. These can only be calculated after the dims are created.
```
all of the optional keyword arguments need an `optional` tag, e.g.
```
    attrs_from_scratch : bool, optional
        ...
```

(see numpydoc HOWTO: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

rabernat commented on xgcm/autogenerate.py:
```diff
@@ -0,0 +1,272 @@
+            is_valued = None
+    else:
+        is_valued = in_val
+    return is_valued
```
are these private functions? If so, they need to be renamed with a leading underscore (i.e. `_parse_wrap_pad`)

rabernat commented on xgcm/grid.py:
```diff
@@ -45,7 +45,8 @@ class Axis:
 
-    def __init__(self, ds, axis_name, periodic=True, default_shifts={}):
+    def __init__(self, ds, axis_name, periodic=True, default_shifts={},
+                 wrap=None):
```
the `wrap` option needs to be added to the docstring for the class.

What does `wrap` do? How does it relate to the `periodic` option?

rabernat commented:
I am worried there is some redundant treatment of periodic coordinates.

Before this PR, "wrapping" of coordinates was already supported via the `periodic` options. I don't understand what the new `wrap` keyword does in the context of `Grid` and `Axis` objects.

jbusecke commented:
Coordinate values are discontinuous, even if the boundary is periodic. So in order to interpolate/diff/etc. a tracer field across a periodic boundary it is enough to 'know' the position of the neighboring grid cell (this is done in xgcm already)

For the coordinate values themselves this is not sufficient: If one wants to interpolate the longitude value between the first and the last longitude cell (e.g. lon1= 0 and lon2=359) one ends up with 179.5 which is wrong. Depending on the 'direction' of interpolation the expected result should be 359.5 (achieved by adding the 'wrap' value 360 to lon1) or 0.5 (achieved by subtracting the same value from lon2). This functionality is added with the 'wrap' keyword and the private wrap function.

So the 'wrap' parameter is defining a subset of a periodic boundary, where values show a well-defined discontinuity across said boundary.

I am not sure if there are other use cases for this, but I would argue that it is simple enough and extends the generality of the gcm axis. 

Does that make things clearer? Maybe we change the name to periodic_discontinuity?

rabernat commented:
Thanks for the explanation. But I still don't understand why `Axis` needs to know about this. As far as I can see, it only comes up when interpolating coordinates. How does `Axis.interp` know whether the variable it is interpolating is a coordinate that needs to be wrapped or a regular variable that does not?

> Maybe we change the name to periodic_discontinuity?
👍  What about `coordinate_discontinuity`? "Wrap" is just too similar to "periodic."

jbusecke commented:
> What about coordinate_discontinuity? "Wrap" is just too similar to "periodic."
I can do that but I think coordinate is too explicit...maybe just boundary_discontinuity?

The reason i put it in axis is so that it is applied on the level of picking the neighboring points (which does most of the heavy lifting in terms of logic). This ensures that the discontinuity is respected for all functions in axis (diff, interp, cumsum, and potential future ones). I chose this over the option to put wrap in each of the functions for simplicity, since the 'deal with discontinuity' function needs to know the direction of the shift (as explained in the above example).

...but I see now how that is a problem if you want to use the axis object on different fields. 
Let me attempt to transfer that into the function arguments instead of the axis init arguments...

jbusecke commented:
Ok this seems more useful now. This way people can run `xgcm.Axis.interp(...., boundary_discontinuity=360)` and `xgcm.Axis.diff(...., boundary_discontinuity=360)`, while still using the same axis for the default behaviour (contious fields across boundary). I did not attempt to add it to cumsum.

jbusecke commented:
Perhaps you could add a test case for the new keyword once everything else is in order? I am still not able to completely decipher you test cases :-)

rabernat commented:
I like this much more! I think this is just about ready to merge!

rabernat commented on xgcm/test/test_autogenerate.py:
```diff
@@ -0,0 +1,286 @@
+        _position_to_relative('left', 'right')
+
+
+def test__():
```
don't you need a better name for this test function?

rabernat commented:
OMG I'm so excited!

jbusecke commented:
Uhoh. Found an error in the documentation. How do I best go about fixing that? New small PR?

rabernat commented:
> New small PR?

Yes. I assume there will be other issues that come up, so no huge rush.

jbusecke commented:
Ok I can work on it while advancing #73.

rabernat commented:
This PR closes #73.
